pepr/utilities.py

```diff
@@ -51,3 +55,270 @@ def assign_record_ids_to_target_models(
             )
 
     return records_per_target_model + offset
+
+
+def filter_out_outlier(
+    data, labels, filter_pars, data_conf, save_path=None, load_pars=None
+):
+    """
+    Filter out potentially vulnerable samples.
+
+    The first 7 steps are identical to the first 7 of the direct GMIA attack
+    (:meth:`pepr.privacy.gmia.DirectGmia`) to find potential vulnerable records.
+
+    Steps:
+
+    1. Create mapping of records to reference models.
+    2. Train the reference models.
+    3. Generate intermediate models.
+    4. Extract reference high-level features.
+    5. Extract target high-level features.
+    6. Compute pairwise distances between reference and target high-level features.
+    7. Determine potential vulnerable target records.
+    8. Remove potential vulnerable target records.
+
+    Parameters
+    ----------
+    data : numpy.ndarray
+        Dataset with all training samples used in the given pentesting setting.
+    labels : numpy.ndarray
+        Array of all labels used in the given pentesting setting.\
+    filter_pars : dict
+        Dictionary containing needed filter parameters:
+
+        * number_reference_models (int): Number of reference models to be trained.
+        * reference_training_set_size (int): Size of the trainings set for each
+          reference model.
+        * create_compile_model (function): Function that returns a compiled
+          TensorFlow model (typically identical to the target model) used in the
+          training of the reference models.
+        * reference_epochs (int): Number of training epochs of the reference models.
+        * reference_batch_size (int): Batch size used in the training of the
+          reference models.
+        * hlf_metric (str): Metric (typically 'cosine') used for the distance
+          calculations in the high-level feature space. For valid metrics see
+          documentation of sklearn.metrics.pairwise_distances.
+        * hlf_layer_number (int): If value is n, the n-th layer of the model
+          returned by create_compile_model is used to extract the high-level feature
+          vectors.
+        * neighbor_threshold (float): If distance is smaller then the neighbor
+          threshold the record is selected as target record.
+        * probability_threshold (float): For details see section 4.3 from the paper.
+        * number_target_records (int): If set, the selection algorithm performs
+          `max_search_rounds`, to find a `neighbor_threshold`, that leads to a finding
+          of `n_targets` target records. These target records are most vulnerable with
+          respect to our selection criterion.
+        * max_search_rounds (int): If `number_target_records` is given, maximal
+          `max_search_rounds` are performed to find `number_target_records` of potential
+          vulnerable target records.
```

Comment:
I think ``number_classes``is missed here? :)

Reply:
Fixed in c3a67ed